### PR TITLE
[tt-train] Centralize random test data generation

### PR DESCRIPTION
**Pull request by mdragulaTT**
Consolidate duplicated random-data generation across `tt-train` tests and benchmarks into shared test utilities.

### Ticket
N/A

### Problem description
Random test/benchmark data generation in `tt-train` was duplicated across many files (custom `parallel_generate` lambdas, local wrappers, `xt::random`, and manual vector copy/adapt code), making maintenance harder and behavior less consistent.

### What's changed
- Added shared helpers in `tt-train/sources/ttml/test_utils/random_data.hpp` and migrated tests/benchmarks to use them.
- Simplified many callsites from `xt::empty + fill_uniform` (or vector+adapt/copy) to direct `make_uniform_xarray` / `make_uniform_vector`.
- Removed unused normal-distribution helpers and redundant local wrappers.
- Renamed `tt-train/tests/ops/layer_norm_composite_test.cpp` to `tt-train/tests/ops/layernorm_composite_test.cpp` and updated `tt-train/tests/CMakeLists.txt`.
- Applied follow-up fixes required for successful build:
  - corrected xtensor include paths in `random_data.hpp`
  - fixed template `static_assert` placement in benchmark/serialization helpers
  - resolved duplicate-symbol issue in cross-entropy tests

### Validation
- `./build_metal.sh -b Release --build-tt-train` passed
- `ctest --test-dir build_Release/tt-train/tests --output-on-failure` passed (`475` passed, `0` failed, `2` disabled)
- Benchmarks executed successfully:
  - `adamw_benchmark`
  - `from_vector_benchmark`
  - `matmuls_benchmark`
  - `swiglu_fusion_benchmark`

### CI Status
_Auto-generated on every push. Badges update live. Click a pipeline name or badge to filter by this branch._

| Pipeline | Status | Latest Run |
|---|:---:|---|
| [Sanity tests](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:mdragula/random-data-utils-tt-train) | [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:mdragula/random-data-utils-tt-train) | [#2111](https://github.com/tenstorrent/tt-metal/actions/runs/24082605328) |
| [Blackhole post-commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:mdragula/random-data-utils-tt-train) | [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:mdragula/random-data-utils-tt-train) | [#17161](https://github.com/tenstorrent/tt-metal/actions/runs/24082625455) |
| [L2 Nightly](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:mdragula/random-data-utils-tt-train) | [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:mdragula/random-data-utils-tt-train) | [#4953](https://github.com/tenstorrent/tt-metal/actions/runs/24082657724) |
| [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mdragula/random-data-utils-tt-train) | [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mdragula/random-data-utils-tt-train) | [Launch ↗](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mdragula/random-data-utils-tt-train) |
| [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mdragula/random-data-utils-tt-train) | [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mdragula/random-data-utils-tt-train) | [Launch ↗](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mdragula/random-data-utils-tt-train) |
| [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mdragula/random-data-utils-tt-train) | [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mdragula/random-data-utils-tt-train) | [Launch ↗](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mdragula/random-data-utils-tt-train) |

### CODE REVIEW

File: tt-train/sources/ttml/test_utils/random_data.hpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+template <typename T>
+inline void fill_uniform(std::span<T> data, T min, T max, uint32_t seed) {
+    if constexpr (std::is_integral_v<T>) {
+        ttml::core::parallel_generate(data, [min, max]() { return std::uniform_int_distribution<T>(min, max); }, seed);
+    } else {
+        const float min_f = static_cast<float>(min);
+        const float max_f = static_cast<float>(max);
+        ttml::core::parallel_generate(
+            data, [min_f, max_f]() { return std::uniform_real_distribution<float>(min_f, max_f); }, seed);
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply tt-train/sources/ttml/test_utils/random_data.hpp]: Makes sense

File: tt-train/tests/ops/layernorm_bw_fused_op_test.cpp [hunk 2]
```diff
@@ -114,26 +115,22 @@ static void CompareKernelVsXArray(
...
-        xt::xarray<float> x_data = xt::empty<float>({total_elements});
         auto rng = autograd::ctx().get_generator();
         uint32_t seed1 = rng();
-        core::parallel_generate<float>(
-            x_data, []() { return std::uniform_real_distribution<float>(-1.0F, 1.0F); }, seed1);
+        xt::xarray<float> x_data =
+            test_utils::make_uniform_xarray<float>(std::array<std::size_t, 1>{total_elements}, seed1, -1.0F, 1.0F);
 
-        xt::xarray<float> gamma_data = xt::empty<float>({features});
         uint32_t seed2 = rng();
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply tt-train/tests/ops/layernorm_bw_fused_op_test.cpp]: Done.

File: tt-train/tests/ops/layernorm_fw_fused_op_test.cpp [hunk 3]
```diff
@@ -70,21 +71,18 @@ static void CompareKernelVsXArray(
         uint32_t total_elements = batch_size * seq_len * heads * features;
         uint32_t combined_batch = batch_size * seq_len * heads;
 
-        xt::xarray<float> x_data = xt::empty<float>({total_elements});
         auto rng = autograd::ctx().get_generator();
         uint32_t seed1 = rng();
-        core::parallel_generate<float>(
-            x_data, []() { return std::uniform_real_distribution<float>(-1.0F, 1.0F); }, seed1);
+        xt::xarray<float> x_data =
+            test_utils::make_uniform_xarray<float>(std::array<std::size_t, 1>{total_elements}, seed1, -1.0F, 1.0F);
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply tt-train/tests/ops/layernorm_fw_fused_op_test.cpp]: Done.

File: tt-train/tests/serialization/flatbuffer_serializer_test.cpp
Comment (vtsilytskyiTT) [inline tt-train/tests/serialization/flatbuffer_serializer_test.cpp:15]: Do we still need <random> here?

File: tt-train/sources/ttml/test_utils/random_data.hpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+template <typename T>
+inline std::vector<T> make_uniform_vector(std::size_t count, T min, T max, uint32_t seed) {
+    std::vector<T> data(count);
+    fill_uniform<T>(std::span<T>{data.data(), data.size()}, min, max, seed);
+    return data;
+}
+
+template <typename T, typename Shape>
+inline xt::xarray<T> make_uniform_xarray(const Shape& shape, uint32_t seed, T min = T{-1}, T max = T{1}) {
```
Comment (vtsilytskyiTT) [inline tt-train/sources/ttml/test_utils/random_data.hpp]: array has default values for min and max, vector doesn't have. Is this intended?

**Review by vtsilytskyiTT — APPROVED** [review]
LGTM, 2 minor comments.

File: tt-train/tests/serialization/flatbuffer_serializer_test.cpp
Comment (mdragulaTT) [reply tt-train/tests/serialization/flatbuffer_serializer_test.cpp:15]: Yes, we still have this functions that uses `std::mt19937`:

```
std::string generate_unique_temp_dir_name() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 15);

    constexpr int name_length = 16;
    std::string name = "flatbuffer_test_";

    for (int i = 0; i < name_length; ++i) {
        name += "0123456789abcdef"[dis(gen)];
    }

    return name;
}
```

File: tt-train/sources/ttml/test_utils/random_data.hpp (hunk 4 above)
Comment (mdragulaTT) [reply tt-train/sources/ttml/test_utils/random_data.hpp]: Fair. Removed default args for consistency - better to have users explicitly specify the range.

**Review by zbaczewskiTT — APPROVED** [review]
LGTM
Thank you!

**Review by jmalone-tt — APPROVED** [review]

File: tt-train/sources/ttml/test_utils/random_data.hpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+#include "core/random.hpp"
+#include "xtensor/containers/xarray.hpp"
+#include "xtensor/generators/xbuilder.hpp"
+
+namespace ttml::test_utils {
+
+template <typename T>
+inline void fill_uniform(std::span<T> data, const T min, const T max, const uint32_t seed) {
+    if constexpr (std::is_integral_v<T>) {
+        ttml::core::parallel_generate(data, [min, max]() { return std::uniform_int_distribution<T>(min, max); }, seed);
```
Comment (philei-tt) [inline tt-train/sources/ttml/test_utils/random_data.hpp:23]: Wan't there some issue with parallel generation, that we had to revert it? Is this fixed?

File: tt-train/tests/benchmark/adamw_benchmark.cpp [hunk 6]
```diff
@@ -76,16 +52,23 @@ void BM_AdamW(benchmark::State& state) {
     const auto dtype = ttnn::DataType::BFLOAT16;
     const ttnn::Shape shape(adamw_shape.shape);
     const uint32_t seed = static_cast<uint32_t>(std::hash<std::string>{}(adamw_shape.name));
+    const auto tensor_spec = ttnn::TensorSpec(
+        shape, tt::tt_metal::TensorLayout(dtype, tt::tt_metal::Layout::TILE, ttnn::DRAM_MEMORY_CONFIG));
+
+    const auto make_tensor = [&](float min, float max, uint32_t tensor_seed) {
+        auto data = ttml::test_utils::make_uniform_vector<float>(shape.volume(), min, max, tensor_seed);
+        return ttnn::Tensor::from_vector(data, tensor_spec, device.get());
+    };
```
Comment (philei-tt) [inline tt-train/tests/benchmark/adamw_benchmark.cpp]: (nit) I would keep make_random_tensor name here. It's more descriptive

**Review by philei-tt — APPROVED** [review]
Please make sure tests pass before merging

File: tt-train/sources/ttml/test_utils/random_data.hpp (hunk 5 above)
Comment (mdragulaTT) [reply tt-train/sources/ttml/test_utils/random_data.hpp:23]: Hmmmm... good question. The revert of #25654 was for an earlier parallel RNG implementation that broke tests (#26431).

Later, RNG was reworked by @athompsonTT via a separate SIMD/legacy path (#30581), with SIMD default briefly enabled (#33500), then reverted (34060880c83), and finally set to opt-in only (#35144).

So this PR does not force SIMD RNG: parallel_generate uses legacy by default, and SIMD only when TT_TRAIN_ENABLE_SIMD_RNG is explicitly set.

I think this fell off the radar after Adam left, so I opened a follow-up issue (https://github.com/tenstorrent/tt-metal/issues/41568) to revisit SIMD RNG. It’s opt-in today due to loss regression, but maybe there is a low-effort fix that preserves correctness and restores perf.

File: tt-train/tests/benchmark/adamw_benchmark.cpp (hunk 6 above)
Comment (mdragulaTT) [reply tt-train/tests/benchmark/adamw_benchmark.cpp]: Sure. Done.

File: tt-train/tests/ops/layernorm_composite_test.cpp [hunk 7]
```diff
@@ -27,27 +28,15 @@ class LayerNormOpTest : public ::testing::Test {
...
+    auto input_data = ttml::test_utils::make_uniform_xarray<float>(
+        std::array<std::size_t, 4>{batch_size, seq_len, heads, features}, -1.0F, 1.0F, 42U);
 
-    std::vector<float> test_data;
-    test_data.reserve((size_t)batch_size * seq_len * heads * features);
-    for (uint32_t i = 0; i < batch_size * seq_len * heads; i++) {
-        float mean = (float)i / (float)size;
-        float stddev = 1.F + (float)i / (float)(size * 2);
-        std::mt19937 gen(i);
-        std::normal_distribution<float> dist(mean, stddev);
```
Comment (imichalakTT) [inline tt-train/tests/ops/layernorm_composite_test.cpp:43]: I guess normal vs uniform doesn't matter here, but we can consider adding normal variants inside test_utils, i.e. `ttml::test_utils::make_normal_xarray`.

File: tt-train/tests/optimizers/muon_composite_test.cpp [hunk 8]
```diff
@@ -94,8 +94,8 @@ TEST_P(MuonCorrectnessTest, DeviceMatchesCPU) {
     using namespace ttml;
     const auto& tc = GetParam();
 
-    xt::xarray<float> w0 = xt::random::randn<float>(tc.shape, 0.0f, 1.0f);
-    xt::xarray<float> g0 = xt::random::randn<float>(tc.shape, 0.0f, 1.0f);
+    xt::xarray<float> w0 = ttml::test_utils::make_uniform_xarray<float>(tc.shape, -1.0F, 1.0F, 42U);
+    xt::xarray<float> g0 = ttml::test_utils::make_uniform_xarray<float>(tc.shape, -1.0F, 1.0F, 43U);
```
Comment (imichalakTT) [inline tt-train/tests/optimizers/muon_composite_test.cpp:98]: Same thing as above

**Review by imichalakTT — APPROVED** [review]

File: tt-train/tests/ops/layernorm_composite_test.cpp (hunk 7 above)
Comment (mdragulaTT) [reply tt-train/tests/ops/layernorm_composite_test.cpp:43]: I discussed this with @dyurshevichTT and @zbaczewskiTT - the normal distributions were added here without any specific reason. We agreed that uniform distribution is sufficient for our testing needs and there's no point in keeping normal. We'll stick with uniform for consistency, but can add normal variants to test_utils if we encounter a specific use case that requires it.

File: tt-train/tests/optimizers/muon_composite_test.cpp (hunk 8 above)
Comment (mdragulaTT) [reply tt-train/tests/optimizers/muon_composite_test.cpp:98]: Answered here: https://github.com/tenstorrent/tt-metal/pull/40910#discussion_r3045441338
